go-common.mk

```diff
@@ -108,15 +102,27 @@ endif
 
 post-clean::
 
+# Generate code
+generate:: pre-generate standard-generate post-generate
+
+pre-generate::
+
+standard-generate::
+ifdef GORUNGENERATE
+	$(GO) generate ./...
+endif # GORUNGENERATE
+
+post-generate::
+
 # Lint code
 lint:: pre-lint standard-lint post-lint
 
 pre-lint::
 
 standard-lint::
-	@which golangci-lint > /dev/null 2>&1 || \
-		$(GO) install github.com/golangci/golangci-lint/cmd/golangci-lint@latest
-	golangci-lint run $(GOLINTFLAGS)
+	go mod why github.com/golangci/golangci-lint/cmd/golangci-lint 2> /dev/null | ( ! grep "does not need package" > /dev/null 2>&1 ) || \
+		(echo "!!!!! golangci-lint is not installed in your go module - please add it in a suitable tools.go file. For example, see here: https://github.com/AchievementNetwork/quiz-api/blob/main/tools.go"; exit 1)
+	go run github.com/golangci/golangci-lint/cmd/golangci-lint run $(GOLINTFLAGS)
```

Comment:
A couple of concerns here.  One is that I don't see any mention of golangci in tools.go in quiz-api or in its go.mod.  Am I missing something?  The other is that this is a good idea in theory, but in practice it will mean manually intervening in all of our Go projects to pick up the new version or we'll be seeing lint failures in all of the builds.  What can we do to prevent that?  I think a transmogrifier recipe could be a solution, but wanted to know what your intentions were there?

Reply:
1. It's been a couple months since I opened this PR, but this was addressed in [an earlier exchange we had](https://github.com/AchievementNetwork/common-make/pull/21#discussion_r1450755675). There's a chicken and egg problem since the PR (which I have open already) that adds the linter to `tools.go` is the same one that switches over to using this new common-make, so I have not merged it yet. That link will be accurate very soon, but not until after this current PR is approved and merged.

2. That's absolutely right, every go project needs to be updated. This is one of the updates I'll be making in all go projects as part of IN-517, the other main one being that they should all run `go generate` in Circle and check in the results. So more PRs will be coming, but I'm not going to open the all the other PRs until after I have approval in common-make and go-orb, in case changes need to be made here.

Reply:
(Oh I just realized that that other PR is not actually open for review yet - I'll get that open in just a minute)

Reply:
Sorry for the slow response here.  Thanks for the explanation.  I dug in a little more on this today and I had a concern about the use of `go mod why`, which I think from testing will also pick up indirect dependencies on golangci-lint.  I think we would like direct dependencies here.  I tried out something like this:

```
standard-lint::
	@go list -m -f '{{.Path}} {{.Indirect}}' all 2> /dev/null | \
		grep -v ' true' | \
		grep 'github.com/golangci/golangci-lint/cmd/golangci-lint' 2>&1 > /dev/null || \
		( if [ -f tools.go ]; then \
			eval "$$_common_make_tools_lint_help"; \
		  else \
			eval "$$_common_make_no_tools_lint_help"; \
		  fi; \
		  exit 1 )
	go run github.com/golangci/golangci-lint/cmd/golangci-lint run $(GOLINTFLAGS)
```

and that seemed to work.  The key there being that indirect dependencies (true) are filtered out.  The variables were defined as

```
define __common_make_no_tools_lint_help
cat <<EOF
No dependency on golangci-lint found!

Please create a top level tools.go file with contents like

-
package tools

import _ "github.com/golangci/golangci-lint/cmd/golangci-lint"
-

(the package may need adjusting) and then run 'go mod tidy'.

Then commit the new file and changes to 'go.mod' and 'go.sum'.
EOF
endef
export _common_make_no_tools_lint_help = $(value __common_make_no_tools_lint_help)

define __common_make_tools_lint_help
cat <<EOF
No dependency on golangci-lint found!

Please add an import of "github.com/golangci/golangci-lint/cmd/golangci-lint"
to the tools.go file and then run 'go mod tidy'.

Then commit the changes to the file and changes to 'go.mod' and 'go.sum'.
EOF
endef
export _common_make_tools_lint_help = $(value __common_make_tools_lint_help)
```

Reply:
I haven't tried out those instructions, but I think they're pretty close.  Please feel free to adjust them based on your experience if you decide to use that suggestion

Reply:
Thanks! The messages and bash syntax are much much clearer.

Unfortunately, that command doesn't seem to work. The command you wrote uses the `-m` flag, which only lists modules, not packages, so it will never find the package `github.com/golangci/golangci-lint/cmd/golangci-lint` no matter what the situation in our go.mod.

On the other hand, if we drop the `-m` flag and list packages, we don't actually have the `.Indirect` field to work with - that field is only defined for modules.

Technically, both approaches are imperfect, since we want both the package, and for it to be direct. In practice though, I think either is probably fine, since I don't think it's likely that we would import part of the golangci-lint repo, or that we would import it indirectly. (For instance, stdmirror relying on go-service, which in turn lists golangci-lint as a _tools_ dependency, doesn't actually result in golangci-lint being included as an indirect dependency of stdmirror. The build constraints in the tools file help to exclude that.)

Reply:
Incidentally, I'm not able to find any way to make `go list` include dependencies mentioned in a file with build constraints. I notice that you removed the build constraints from the tools.go file, which I suspect is key to why you were able to use `go list`. But the build constraints are not optional - in general, importing an executable package in go file that isn't excluded by build constraints will cause a compiler error, especially for those projects that have non-tools go code in the same folder.

Reply:
Thanks Xander.  I experimented a little with this today and also couldn't see a way to either get `go list` to do what we want or to get `go mod why` to output in another format.

The one thing that I would say that might be a minor improvement is that the `go mod why` docs (https://go.dev/ref/mod#go-mod-why) don't specify the format of the error message (which we're relying on), other than it being in parentheses, but do specify a couple of other things that might be useful.  Of those, if the package *is* present as a dependency then it will display the path through the graph to that package.  I.e., the last line of the output is going to be the package we're looking for.  Given that, we could pipe the output through `tail -1` and assert that the resulting string is equal to `github.com/golangci/golangci-lint/cmd/golangci-lint` and output the error if that isn't true.  Alternatively we could look for a line that beings and ends in parentheses as meaning it isn't present as well.

That's minor though, so I'll leave it up to you

Reply:
Hmm this is tricky because you're right that the parentheses seem to be the only guarantee about that message from the docs, but they're also less meaningful and so makes the logic much more opaque IMO. I think your other suggestion - to reverse the logic and check for the path on the last line of the success output, seems like the best option. Thanks for looking into that! I checked for a structured format and then quit when I saw there wasn't one :)